client/src/main/java/com/battleships/connection/Connection.java

```diff
@@ -0,0 +1,209 @@
+    private final BooleanProperty playerActive = new SimpleBooleanProperty(false);
+    private final BooleanProperty playerReadyProperty = new SimpleBooleanProperty(false);
+
+    public boolean getPlayerActive() {
+        return playerActive.get();
+    }
+
+    public BooleanProperty playerActiveProperty() {
+        return playerActive;
+    }
+
+    public void setPlayerActive(boolean playerActive) {
+        this.playerActive.set(playerActive);
+    }
+
+    public void setPlayerReady(boolean playerReady) {
+        this.playerReadyProperty.setValue(playerReady);
+    }
+
+    public BooleanProperty playerReadyProperty() {
+        return this.playerReadyProperty;
+    }
+
+    /**
+     * Establishes the server's connection
+     *
+     * @param connectionInfo - <code>ConnectionInfo</code> object that includes the ip and port of the server
+     */
+
+    public void establishConnection(ConnectionInfo connectionInfo) {
+        if (!isConnected()) {
+            tryToEstablishConnection(connectionInfo);
+        }
+    }
+
+    public BooleanProperty connectedProperty() {
+        return connected;
+    }
+
+    private void setConnected(boolean connected) {
+        this.connected.set(connected);
+    }
+
+    public boolean isConnected() {
+        checkConnected();
+        return connected.get();
+    }
+
+    private void checkConnected() {
+        setConnected(socket != null && socket.isConnected());
+    }
+
+    private void tryToEstablishConnection(ConnectionInfo connectionInfo) {
+        try {
+            socket = new Socket();
+            InetSocketAddress endpoint = new InetSocketAddress(connectionInfo.getIp(), connectionInfo.getPort());
+            socket.connect(endpoint, initialConnectingTimeout);
+            establishServerIO();
+        } catch (IOException e) {
+            String errorMessage = String.format(LogMessages.CANNOT_CONNECT_TO_SERVER, connectionInfo.getIp(), connectionInfo.getPort());
+            logger.error(errorMessage);
+            //TODO playerName to user that problems occurred
+        }
+    }
+
+    /**
+     * Disconnects client from the server
+     */
+
+    public void disconnect() {
+        if (isConnected()) {
+            tryToDisconnectFromServer();
+        }
+    }
+
+    private void tryToDisconnectFromServer() {
+        try {
+            Message<String> message = new Message<>(CommandType.STOP_PLAYING, "");
+            sendToServer(message);
+            disconnectPlayer();
+        } catch (IOException e) {
+            logger.error(LogMessages.PROBLEM_WHEN_TRYING_TO_DISCONNECT + e.getMessage());
+        } finally {
+            socket = null;
+        }
+    }
+
+    /**
+     * Sends to the connected server specified command with command value
+     *
+     * @param message - <code>Message</code> Player command object with specified command kind and value
+     */
+
+    public <V> void sendToServer(Message<V> message) {
```

Comment:
Please wonder about action if condition fails.

Reply:
We will implement that for sure in the future